src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -324,12 +324,16 @@ def plan_state(self) -> Dict[PlanState, Any]:
 
     @property
     def cumulative_node_complexity(self) -> Dict[PlanNodeCategory, int]:
-        if self._cumulative_node_complexity is None:
-            self._cumulative_node_complexity = sum_node_complexities(
-                self.individual_node_complexity,
-                *(node.cumulative_node_complexity for node in self.children_plan_nodes),
-            )
-        return self._cumulative_node_complexity
+        with self.analyzer.session._plan_lock:
```

Comment:
Just curious, is this race condition only happen in select statement or is other statement dose not need special treatment?

Reply:
These methods are part of `Selectables` and `SnowflakePlan` and they are more vulnerable to be lazily triggered together. For most other methods, they are triggered as a part of `analyzer.resolve()` so they are safer.